src/hnsw/validate_index.c

```diff
@@ -0,0 +1,650 @@
+                     blockmap->first_id,
+                     group_node_first_index,
+                     blockmap_id,
+                     HNSW_BLOCKMAP_BLOCKS_PER_PAGE);
+            }
+            HnswIndexPageSpecialBlock *special = (HnswIndexPageSpecialBlock *)PageGetSpecialPointer(page);
+            if(special->firstId != blockmap->first_id) {
+                elog(ERROR,
+                     "special->firstId=%" PRIu32 " != blockmap->first_id=%" PRIu32
+                     " for "
+                     "blockmap_block=%" PRIu32 " blockmap_groupno=%d blockmap_id=%d",
+                     special->firstId,
+                     blockmap->first_id,
+                     blockmap_block,
+                     blockmap_groupno,
+                     blockmap_id);
+            }
+            if(special->lastId != special->firstId + HNSW_BLOCKMAP_BLOCKS_PER_PAGE - 1) {
+                elog(ERROR,
+                     "special->lastId=%" PRIu32 " != (special->first_id=%" PRIu32
+                     " + HNSW_BLOCKMAP_BLOCKS_PER_PAGE=%d - 1) for "
+                     "blockmap_block=%" PRIu32 " blockmap_groupno=%d blockmap_id=%d",
+                     special->lastId,
+                     special->firstId,
+                     HNSW_BLOCKMAP_BLOCKS_PER_PAGE,
+                     blockmap_block,
+                     blockmap_groupno,
+                     blockmap_id);
+            }
+            /* TODO confirm this */
+            /*
+            expected_special_nextblockno = blockmap_id == number_of_blockmaps_in_group - 1 ?
+                                           InvalidBlockNumber : blockmap_block + 1;
+            */
+            expected_special_nextblockno = blockmap_block + 1;
+            if(special->nextblockno != expected_special_nextblockno) {
+                elog(ERROR,
+                     "special->nextblockno=%" PRIu32 " != expected_special_nextblockno=%" PRIu32
+                     " for "
+                     "blockmap_block=%" PRIu32 " blockmap_groupno=%d blockmap_id=%d",
+                     special->nextblockno,
+                     expected_special_nextblockno,
+                     blockmap_block,
+                     blockmap_groupno,
+                     blockmap_id);
+            }
+            ldb_vi_analyze_blockmap(blockmap, vi_blocks, blocks_nr, vi_nodes, nodes_nr);
+
+            UnlockReleaseBuffer(buf);
+        }
+        nodes_remaining -= Min(batch_size, nodes_remaining);
+        group_node_first_index += batch_size;
+        batch_size = batch_size * 2;
+        blockmap_groupno++;
+    }
+}
+
+static bool ldb_vi_read_node_chunk(void *chunk, size_t chunk_size, void *tape, unsigned tape_pos, unsigned tape_size)
+{
+    if(tape_pos + chunk_size > tape_size) return false;
+    memcpy(chunk, (char *)tape + tape_pos, chunk_size);
+    return true;
+}
+
+#define LDB_VI_READ_NODE_CHUNK(_chunk, _tape, _tape_pos, _tape_size)                              \
```

Comment:
To have precise name of whatever is being read in case of a read failure:
```247                  "Error reading " #_chunk```
Also to add a chunk size:
```243         size_t _chunk_size = sizeof(_chunk);```

Alternatives: to pass a separate string parameter (error-prone and duplicate of `_chunk` parameter) or just to tell that "node read has failed" without being specific (worse error reporting, longer time to figure out what's wrong etc.). Without a macro the chunk size has to be passed anyway (which would be an error-prone duplicate of code).

I've changed the macro to be just a thing that adds `#_chunk` and the chunk size to the list of the function parameters. This would significantly simplify the macro with no undesirable side effects. It would even have a benefit that the `tape_pos` would be updated explicitly in the function, not somewhere inside a macro.

Thanks for highlighting a part of the code that could be substantially improved.